### ISSUE TITLE
How do arpack, propack and lobcg compare to each other?

### ISSUE THREAD

**Issue by ERijck** [scipy/scipy#17023]
I want to implement singular value decomposition on a sparse matrix using the Scipy implementation. The [documentation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.svds.html) shows three solvers: ‘arpack’, ‘propack’ and ‘lobpcg’. How do these solvers compare in terms of performance and running time?

**Comment by tupui** [comment]
Hi @ERijck, thank you for reporting. @lobpcg do we have any documentation on that? I did not really find much and it's true that it would be a good thing to have a few guidelines in the doc.

**Comment by lobpcg** [comment]
No, there is no any documentation for three `svds` solvers: `arpack`, `propack`, and `lobpcg` in terms of performance and running time. Yes, there must be. Benchmarking should have been a requirement for the latest addition, `propack` to be merged, but it was not. 

There might be a similar deficiency of benchmarking for `TruncatedSVD` in `sklearn`; see attempts, e.g., related to https://github.com/scikit-learn/scikit-learn/pull/12319, to benchmark: 
- https://github.com/rth/notebooks/blob/master/sklearn/PR_12319/bench_lobpcg_truncated_svd.ipynb
- https://github.com/glemaitre/bench_lobpcg

However,  `svds` in `scipy` needs several fixes before benchmarking  can be meaningfully well done and documented, e.g.,
- https://github.com/scipy/scipy/pull/16712
- https://github.com/scipy/scipy/issues/16330

With the current pace of reviewing/merging, fixing those may take months (may be years :-) 

Before that happens, users should definitely be encouraged to carefully benchmark all 3 solvers available varying the max number of iterations and checking the accuracy manually rather than trusting the essentially broken tolerance in  `svds`. For small-size matrices, dense 'svd' will be evidently competitive and  must be also included in benchmarking.

Finally, `arpack` and `propack` are interfaces to pre-compiled libraries, while `lobpcg` is a pure Python code. On the one hand, it makes `lobpcg` automatically slower (unless the user compiles it in to binary, which surely would be a good idea). On the other hand, that allows  `lobpcg` to natively support pretty much all numeric dtypes, in contrast to  `arpack` and `propack`.

**Comment by tupui** [comment]
Thank you for the write up @lobpcg

**Comment by mdhaber** [comment]
> Benchmarking should have been a requirement for the latest addition, propack to be merged, but it was not.

It was. You can see benchmark results of the three solvers for many problems here.

https://pv.github.io/scipy-bench/#sparse_linalg_svds.BenchSVDS.time_svds?p-solver='arpack'&p-solver='lobpcg'&p-solver='propack'&p-problem='abb313'

You can run the benchmarks in this file:
https://github.com/scipy/scipy/blob/main/benchmarks/benchmarks/sparse_linalg_svds.py

**Comment by lobpcg** [comment]
@mdhaber thanks - great, I did not know!
